### PR TITLE
fix(verify-bytecode): check contract name in cache

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

Fixes an issue where if multiple ABI files were located in the `out/Contract.sol/**` directory it would fetch the first ABI file or the one where the solidity version matched on the first round without matching the `name` of the ABI file with the contract being verified. 

This would lead to fetching the wrong file from cache. 

For example, if the bytecode being verified is for the `PSMVariant1Actions` contract, the `out/PSMVariant1Actions.sol` tree looks like:

```bash
├── PSMVariant1Actions.sol
│   ├── GemJoinLike.json
│   ├── PSMVariant1Actions.json
│   └── PSMVariant1Like.json
```

It would lead to `GemJoinLike.json` being fetched from cache instead of `PSMVariant1Actions.json`

## Solution

Strongly check the `name` of the abi file and then the solidity version.

### CODE REVIEW

**Comment by yash-atreya** [comment]
@mattsse ptal

File: crates/verify/src/bytecode.rs [hunk 1]
```diff
@@ -224,11 +224,14 @@ impl VerifyBytecodeArgs {
                 (VerificationType::Partial, _) => (VerificationType::Partial, true),
             };
 
+        trace!("verification_type: {:?}", verification_type);
+        trace!("has_metadata: {}", has_metadata);
```
Comment (mattsse) [inline crates/verify/src/bytecode.rs:228]: let's combine them into one trace

File: crates/verify/src/bytecode.rs [hunk 2]
```diff
@@ -410,37 +413,45 @@ impl VerifyBytecodeArgs {
...
                 return None;
             }
             // Parse etherscan version string
             let version =
                 version.split('+').next().unwrap_or("").trim_start_matches('v').to_string();
+
+            // Check if `out/directory` name matches the contract name
             if key.ends_with(name.as_str()) {
-                if let Some(artifact) = value.into_iter().next() {
+                // let artifacts: Vec<_> =
```
Comment (mattsse) [inline crates/verify/src/bytecode.rs:426]: rm docs

File: crates/verify/src/bytecode.rs [hunk 3]
```diff
@@ -410,37 +413,45 @@ impl VerifyBytecodeArgs {
...
+                let artifacts =
+                    value.iter().flat_map(|(_, artifacts)| artifacts.iter()).collect::<Vec<_>>();
+
+                for artifact in artifacts {
+                    // Check if ABI file matches the name
+                    if !artifact.file.ends_with(name.replace(".sol", ".json")) {
```
Comment (mattsse) [inline crates/verify/src/bytecode.rs:432]: you can reuse the modified name instead of replacing in a loop

**Review by mattsse — CHANGES_REQUESTED** [review]
a few nits

File: crates/verify/src/bytecode.rs (hunk 1 above)
Comment (DaniPopes) [reply crates/verify/src/bytecode.rs:228]: can also use variables directly: `(?verification_type, has_metadata)`

**Review by mattsse — APPROVED** [review]
pending @DaniPopes

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**